### PR TITLE
Enforce coverage requirement and reporting via codecov

### PR DESCRIPTION
**Pull request by Hook25**
## Description

This upgrades our coverage reporting by uploading to codecov.io. Using the features of said service, this also starts to enforce a patch based coverage requirement (for now just hardcoded at 80%).

## Resolved issues

Resolves: https://warthogs.atlassian.net/browse/CHECKBOX-799

## Documentation

This adds a new section to the contributing guide that explains how to get a local coverage report, why we want a coverage constraint and how we enforce it. This section also points to where the coverage constraint is specified and how it is reported io PRs (the pointer is there because in the future we may want to change the requirement, we don't want to have to update this file with the new one!)

## Tests

N/A

### CODE REVIEW

**Comment by Hook25** [comment]
The codecov message works :) the issue it is pointing at is due to the fact that he does not know how much coverage we have in main yet

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -103,6 +103,40 @@ tests for providers:
...
+```
+$ python -m pip install coverage pytest pytest-cov
+# cd to where your test is
+$ python -m coverage run -m pytest .
+```
+Note that every part of this repository has a `.coveragerc` file, they should
+already include anything you may want to see in the report. If something is
+missing you can edit it but please, consult with the team before doing so.
+Tests are intentionally excluded from the coverage report, this is because
+test files tend to inflationate the coverage with no real benefit, so don't
```
Comment (kissiel) [inline CONTRIBUTING.md:128]: s/inflationate/inflate

File: codecov.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+coverage:
+  status:
+    patch:
+      default:
+        target: 80%
```
Comment (kissiel) [inline codecov.yml:5]: What's the source of this value?

**Review by kissiel — COMMENTED** [review]
One suggestion and one question below.

File: codecov.yml (hunk 2 above)
Comment (Hook25) [reply codecov.yml:5]: uhm seems like the card was edited? pretty sure I took the card suggestion on the coverage and copy-pasted the snippet. I will raise it to 90%

**Review by kissiel — APPROVED** [review]
Looks good!
+1

**Pull request merged**